### PR TITLE
migrate category sync

### PR DESCRIPTION
**Pull request by salander85**
Additional changes:
- Add IT for `UnresolvedReferenceResolverService` and adjustments for the service
- Add model `WaitingToBeResolvedCategories` incl. unittests
- Move utility function `mockGraphQLResponse` to common test package
- Extract helper to create different Exceptions used in IT to `ITUtils`

Follow-up:
- Adjust calculation of updateActions for customFields in `CustomUpdateActionUtils`
- JIRA: https://commercetools.atlassian.net/browse/DEVX-169

JIRA: https://commercetools.atlassian.net/browse/DEVX-83

### CODE REVIEW

File: src/main/java/com/commercetools/sync/sdk2/categories/CategorySync.java [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+      @Nonnull final Category oldCategory,
+      @Nonnull final CategoryDraft newCategory,
+      @Nonnull final List<CategoryUpdateAction> updateActions) {
+    final String categoryKey = oldCategory.getKey();
+    return categoryService
+        .updateCategory(oldCategory, updateActions)
+        .handle(ImmutablePair::new)
+        .thenCompose(
+            updateResponse -> {
+              final Throwable sphereException = updateResponse.getValue();
```
Comment (lojzatran) [inline src/main/java/com/commercetools/sync/sdk2/categories/CategorySync.java:352]:
```suggestion
              final Throwable ctpException = updateResponse.getValue();
```

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/commons/utils/CategoryITUtils.java [hunk 2]
```diff
@@ -254,6 +258,17 @@ public static Type ensureCategoriesCustomType(
         typeKey, locale, name, singletonList(ResourceTypeId.CATEGORY), ctpClient);
   }
 
+  /**
+   * Deletes all categories and types from the CTP project defined by the {@code ctpClient}.
+   *
+   * @param ctpClient defines the CTP project to delete the categories and types from.
+   */
+  public static void deleteCategorySyncTestData(@Nonnull final ProjectApiRoot ctpClient) {
```
[the remark this answers is not in the document]
Comment (salander85) [reply src/integration-test/java/com/commercetools/sync/integration/sdk2/commons/utils/CategoryITUtils.java:266]: [fcd9f16](https://github.com/commercetools/commercetools-sync-java/pull/976/commits/fcd9f162ca16fdd7e82f3b2155b346947864aa1d)

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/ctpprojectsource/categories/CategorySyncIT.java [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+
+    assertThat(syncStatistics).hasValues(3, 1, 2, 0, 0);
+    assertThat(callBackErrorResponses).isEmpty();
+    assertThat(callBackExceptions).isEmpty();
+    assertThat(callBackWarningResponses).isEmpty();
+  }
+
+  @Test
+  void syncDrafts_withNewShuffledBatchOfCategories_ShouldCreateCategories() {
+    // Create a total of 130 categories in the source project
```
[the remark this answers is not in the document]
Comment (salander85) [reply src/integration-test/java/com/commercetools/sync/integration/sdk2/ctpprojectsource/categories/CategorySyncIT.java:159]: Plus 5 for the firstly created ones :) I will adjust the comment.

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/ctpprojectsource/categories/CategorySyncIT.java [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+                    .build())
+            .build();
+
+    CTP_TARGET_CLIENT.categories().create(parentDraft).execute().toCompletableFuture().join();
+
+    final CategoryDraft childDraft =
+        CategoryDraftBuilder.of()
+            .name(LocalizedString.of(Locale.ENGLISH, "child"))
+            .slug(LocalizedString.of(Locale.ENGLISH, "child"))
+            .key("child")
```
[the remark this answers is not in the document]
Comment (salander85) [reply src/integration-test/java/com/commercetools/sync/integration/sdk2/ctpprojectsource/categories/CategorySyncIT.java]: [145fead](https://github.com/commercetools/commercetools-sync-java/pull/976/commits/145fead22a94ace8a3d93f29183cafa0070fb991)

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/ctpprojectsource/categories/CategorySyncIT.java [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+        .allSatisfy(
+            throwable -> {
+              assertThat(throwable).isExactlyInstanceOf(CompletionException.class);
+              assertThat(throwable).hasCauseExactlyInstanceOf(BadRequestException.class);
+              final BadRequestException errorResponse = (BadRequestException) throwable.getCause();
+
+              final List<DuplicateFieldError> fieldErrors =
+                  errorResponse.getErrorResponse().getErrors().stream()
+                      .map(
+                          sphereError -> {
```
[the remark this answers is not in the document]
Comment (salander85) [reply src/integration-test/java/com/commercetools/sync/integration/sdk2/ctpprojectsource/categories/CategorySyncIT.java]: [145fead](https://github.com/commercetools/commercetools-sync-java/pull/976/commits/145fead22a94ace8a3d93f29183cafa0070fb991)

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/ctpprojectsource/categories/CategorySyncIT.java [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+            .getBody()
+            .getResults();
+
+    final List<CategoryDraft> categoryDrafts =
+        CategoryTransformUtils.toCategoryDrafts(
+                CTP_SOURCE_CLIENT, referenceIdToKeyCache, categories)
+            .join();
+    Collections.shuffle(categoryDrafts);
+
+    CategorySync categorySyncWith1BatchSize = new CategorySync(buildCategorySyncOptions(1));
```
[the remark this answers is not in the document]
Comment (salander85) [reply src/integration-test/java/com/commercetools/sync/integration/sdk2/ctpprojectsource/categories/CategorySyncIT.java]: [145fead](https://github.com/commercetools/commercetools-sync-java/pull/976/commits/145fead22a94ace8a3d93f29183cafa0070fb991)

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/ctpprojectsource/categories/CategorySyncIT.java [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+            .getBody()
+            .getResults();
+
+    final List<CategoryDraft> categoryDrafts =
+        CategoryTransformUtils.toCategoryDrafts(
+                CTP_SOURCE_CLIENT, referenceIdToKeyCache, categories)
+            .join();
+    Collections.shuffle(categoryDrafts);
+
+    CategorySync categorySyncWith13BatcheSize = new CategorySync(buildCategorySyncOptions(13));
```
[the remark this answers is not in the document]
Comment (salander85) [reply src/integration-test/java/com/commercetools/sync/integration/sdk2/ctpprojectsource/categories/CategorySyncIT.java]: [145fead](https://github.com/commercetools/commercetools-sync-java/pull/976/commits/145fead22a94ace8a3d93f29183cafa0070fb991)

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/ctpprojectsource/categories/CategorySyncIT.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (salander85) [reply src/integration-test/java/com/commercetools/sync/integration/sdk2/ctpprojectsource/categories/CategorySyncIT.java]: [145fead](https://github.com/commercetools/commercetools-sync-java/pull/976/commits/145fead22a94ace8a3d93f29183cafa0070fb991)

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/ctpprojectsource/categories/CategorySyncIT.java [hunk 8]
```diff
@@ -0,0 +1,507 @@
...
+
+  /** Cleans up the target and source test data that were built in this test class. */
+  @AfterAll
+  static void tearDown() {
+    deleteCategorySyncTestData(CTP_TARGET_CLIENT);
+    deleteCategorySyncTestData(CTP_SOURCE_CLIENT);
+  }
+
+  @Test
+  void syncDrafts_withChangesOnly_ShouldUpdateCategories() {
```
[the remark this answers is not in the document]
Comment (salander85) [reply src/integration-test/java/com/commercetools/sync/integration/sdk2/ctpprojectsource/categories/CategorySyncIT.java]: [145fead](https://github.com/commercetools/commercetools-sync-java/pull/976/commits/145fead22a94ace8a3d93f29183cafa0070fb991)

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/categories/CategorySyncIT.java [hunk 9]
```diff
@@ -0,0 +1,789 @@
...
+            .name(LocalizedString.of(Locale.ENGLISH, "furniture"))
+            .slug(LocalizedString.of(Locale.ENGLISH, "furniture"))
+            .key(oldCategoryKey)
+            .build();
+
+    final CategorySyncStatistics syncStatistics =
+        categorySync.sync(singletonList(categoryDraft)).toCompletableFuture().join();
+
+    assertThat(syncStatistics).hasValues(1, 0, 1, 0, 0);
+  }
```
[the remark this answers is not in the document]
Comment (salander85) [reply src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/categories/CategorySyncIT.java]: [094c100](https://github.com/commercetools/commercetools-sync-java/pull/976/commits/094c1001f341909b0aff658cf3bc96f187ef854b)

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/categories/CategorySyncIT.java [hunk 10]
```diff
@@ -0,0 +1,789 @@
...
+                  final ApiHttpMethod method = request.getMethod();
+                  if (uri.contains("categories/key=" + categoryKey)
+                      && ApiHttpMethod.GET.equals(method)) {
+                    if (this.requestInvocationCounter > 0) {
+                      return CompletableFutureUtils.exceptionallyCompletedFuture(
+                          createBadGatewayException());
+                    }
+                    this.requestInvocationCounter++;
+                  }
+                  if (uri.contains("categories/" + this.oldCategory.getId())
```
[the remark this answers is not in the document]
Comment (salander85) [reply src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/categories/CategorySyncIT.java]: [094c100](https://github.com/commercetools/commercetools-sync-java/pull/976/commits/094c1001f341909b0aff658cf3bc96f187ef854b)

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/categories/CategorySyncIT.java [hunk 11]
```diff
@@ -0,0 +1,789 @@
...
+  @Nonnull
+  private ProjectApiRoot buildClientWithConcurrentModificationUpdate() {
+    final ProjectApiRoot testClient =
+        ApiRootBuilder.of(
+                request -> {
+                  final String uri = request.getUri() != null ? request.getUri().toString() : "";
+                  final ApiHttpMethod method = request.getMethod();
+                  if (uri.contains("categories/" + this.oldCategory.getId())
+                      && ApiHttpMethod.POST.equals(method)) {
+                    if (this.requestInvocationCounter == 0) {
```
[the remark this answers is not in the document]
Comment (salander85) [reply src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/categories/CategorySyncIT.java]: [094c100](https://github.com/commercetools/commercetools-sync-java/pull/976/commits/094c1001f341909b0aff658cf3bc96f187ef854b)

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/categories/CategorySyncIT.java [hunk 12]
```diff
@@ -0,0 +1,789 @@
...
+    final CategorySyncStatistics syncStatistics =
+        categorySync.sync(singletonList(categoryDraft)).toCompletableFuture().join();
+
+    assertThat(syncStatistics).hasValues(1, 0, 1, 0, 0);
+  }
+
+  @Test
+  void syncDrafts_WithConcurrentModificationException_ShouldRetryToUpdateNewCategoryWithSuccess() {
+    // Preparation
+    final ProjectApiRoot spyClient = buildClientWithConcurrentModificationUpdate();
```
[the remark this answers is not in the document]
Comment (salander85) [reply src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/categories/CategorySyncIT.java]: [094c100](https://github.com/commercetools/commercetools-sync-java/pull/976/commits/094c1001f341909b0aff658cf3bc96f187ef854b)

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/categories/CategorySyncIT.java [hunk 13]
```diff
@@ -0,0 +1,789 @@
...
+            .build();
+
+    final CategorySyncStatistics syncStatistics =
+        categorySync.sync(singletonList(categoryDraft)).toCompletableFuture().join();
+
+    assertThat(syncStatistics).hasValues(1, 0, 0, 0, 0);
+  }
+
+  @Test
+  void syncDrafts_WithChangedCategory_ShouldUpdateCategory() {
```
[the remark this answers is not in the document]
Comment (salander85) [reply src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/categories/CategorySyncIT.java]: [094c100](https://github.com/commercetools/commercetools-sync-java/pull/976/commits/094c1001f341909b0aff658cf3bc96f187ef854b)

File: src/test/java/com/commercetools/sync/sdk2/categories/CategorySyncTest.java [hunk 14]
```diff
@@ -0,0 +1,813 @@
...
+  private List<Throwable> errorCallBackExceptions;
+  private UnresolvedReferencesService mockUnresolvedReferencesService;
+
+  // protected method access helper
+  private static class CategorySyncMock extends CategorySync {
+    CategorySyncMock(
+        @Nonnull final CategorySyncOptions syncOptions,
+        @Nonnull final TypeService typeService,
+        @Nonnull final CategoryService categoryService,
+        @Nonnull final UnresolvedReferencesService unresolvedReferencesService) {
```
[the remark this answers is not in the document]
Comment (salander85) [reply src/test/java/com/commercetools/sync/sdk2/categories/CategorySyncTest.java]: [c61ccfb](https://github.com/commercetools/commercetools-sync-java/pull/976/commits/c61ccfb3eb6ca0bf3064d88b6662aeaaca1fa9fb)

File: src/test/java/com/commercetools/sync/sdk2/categories/CategorySyncTest.java [hunk 15]
```diff
@@ -0,0 +1,813 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+
+@SuppressWarnings("unchecked")
+class CategorySyncTest {
+  private CategorySyncOptions categorySyncOptions;
+  private List<String> errorCallBackMessages;
+  private List<Throwable> errorCallBackExceptions;
+  private UnresolvedReferencesService mockUnresolvedReferencesService;
```
[the remark this answers is not in the document]
Comment (salander85) [reply src/test/java/com/commercetools/sync/sdk2/categories/CategorySyncTest.java]: [c61ccfb](https://github.com/commercetools/commercetools-sync-java/pull/976/commits/c61ccfb3eb6ca0bf3064d88b6662aeaaca1fa9fb)

File: src/main/java/com/commercetools/sync/sdk2/categories/CategorySync.java (hunk 1 above)
Comment (salander85) [reply src/main/java/com/commercetools/sync/sdk2/categories/CategorySync.java]: [c61ccfb](https://github.com/commercetools/commercetools-sync-java/pull/976/commits/c61ccfb3eb6ca0bf3064d88b6662aeaaca1fa9fb)

**Review by lojzatran — APPROVED** [review]

**Pull request merged**